### PR TITLE
Modernized code with clang-tidy [PCL-2731]

### PR DESCRIPTION
**Pull request by gnawme**
Modernization of code base using `clang-tidy` and an innocuous subset of its `modernize` checks:

- [ ] modernize-avoid-bind
- [ ] modernize-avoid-c-arrays
- [ ] modernize-concat-nested-namespaces
- [x] modernize-deprecated-headers
- [ ] modernize-deprecated-ios-base-aliases
- [ ] modernize-loop-convert
- [ ] modernize-make-shared
- [ ] modernize-make-unique
- [ ] modernize-pass-by-value
- [ ] modernize-raw-string-literal
- [x] modernize-redundant-void-arg
- [ ] modernize-replace-auto-ptr
- [ ] modernize-replace-random-shuffle
- [ ] modernize-return-braced-init-list
- [ ] modernize-shrink-to-fit
- [ ] modernize-unary-static-assert
- [ ] modernize-use-auto
- [ ] modernize-use-bool-literals
- [x] modernize-use-default-member-init
- [ ] modernize-use-emplace
- [x] modernize-use-equals-default
- [x] modernize-use-equals-delete
- [ ] modernize-use-nodiscard
- [ ] modernize-use-noexcept
- [x] modernize-use-nullptr
- [x] modernize-use-override
- [ ] modernize-use-trailing-return-type
- [ ] modernize-use-transparent-functors
- [ ] modernize-use-uncaught-exceptions
- [x] modernize-use-using

This PR splits off from #4560; per @kunaltyagi, it enables `clang-tidy` for the entire project, but omits the GitHub Workflow additions that would run `clang-tidy` checks in CI.

### CODE REVIEW

File: features/include/pcl/features/organized_edge_detection.h [hunk 1]
```diff
@@ -82,9 +82,8 @@ namespace pcl
 
       /** \brief Destructor for OrganizedEdgeBase */
 
-      ~OrganizedEdgeBase ()
-      {
-      }
+      ~OrganizedEdgeBase () override
```
[the remark this answers is not in the document]
Comment (gnawme) [reply features/include/pcl/features/organized_edge_detection.h]: Yes, we'll see if the formatting pass in CI objects...

File: features/include/pcl/features/statistical_multiscale_interest_region_extraction.h [hunk 2]
```diff
@@ -71,7 +71,7 @@ namespace pcl
 
       /** \brief Empty constructor */
       StatisticalMultiscaleInterestRegionExtraction ()
-      {};
+      = default;;
```
[the remark this answers is not in the document]
Comment (gnawme) [reply features/include/pcl/features/statistical_multiscale_interest_region_extraction.h]: Same as above, I'll do it and see whether it gets through CI's formatting check.

File: gpu/kinfu/src/cuda/ray_caster.cu [hunk 3]
```diff
@@ -273,7 +273,7 @@ namespace pcl
             break;
           }
 
-        }          /* for(;;)  */
+        }          /* for(;)  */
```
Comment (kunaltyagi) [inline gpu/kinfu/src/cuda/ray_caster.cu:276]: That's not how `;;` works, but I think it'd be better to remove the comment here

File: filters/include/pcl/filters/voxel_grid.h [hunk 4]
```diff
@@ -210,9 +210,8 @@ namespace pcl
       }
 
       /** \brief Destructor. */
-      ~VoxelGrid ()
-      {
-      }
+      ~VoxelGrid () override
+      = default;
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/voxel_grid.h:214]:
```suggestion
      ~VoxelGrid () override = default;
```

File: filters/include/pcl/filters/voxel_grid.h [hunk 5]
```diff
@@ -536,9 +535,8 @@ namespace pcl
...
```
Comment (kunaltyagi) [inline filters/include/pcl/filters/voxel_grid.h:539]:
```suggestion
      ~VoxelGrid () override = default;
```

File: io/include/pcl/compression/color_coding.h [hunk 6]
```diff
@@ -72,8 +72,7 @@ class ColorCoding
   /** \brief Empty class constructor. */
   virtual
   ~ColorCoding ()
-  {
-  }
+  = default;
```
Comment (kunaltyagi) [inline io/include/pcl/compression/color_coding.h:75]: Other places also have the `= default` split on 2 lines

File: recognition/include/pcl/recognition/dot_modality.h [hunk 7]
```diff
@@ -48,7 +48,7 @@ namespace pcl
   {
     public:
 
-      virtual ~DOTModality () {};
+      virtual ~DOTModality () = default;;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/dot_modality.h:51]:
```suggestion
      virtual ~DOTModality () = default;
```

**Review by kunaltyagi — CHANGES_REQUESTED** [review]
156/309 reviewed

File: gpu/kinfu/src/cuda/ray_caster.cu (hunk 3 above)
Comment (gnawme) [reply gpu/kinfu/src/cuda/ray_caster.cu]: I was trying to get rid of double '::', I think I fat-fingered 'Replace' in this instance

File: io/include/pcl/compression/color_coding.h (hunk 6 above)
Comment (gnawme) [reply io/include/pcl/compression/color_coding.h:75]: Caught a few more, hope they're the last

File: recognition/include/pcl/recognition/dot_modality.h (hunk 7 above)
Comment (gnawme) [reply recognition/include/pcl/recognition/dot_modality.h]: Caught a few more of these

**Review by kunaltyagi — CHANGES_REQUESTED** [review]
There's still plenty left (split default).

You can catch them all using `git diff master | grep '^ \+= default'

File: surface/include/pcl/surface/bilateral_upsampling.h [hunk 8]
```diff
@@ -64,7 +64,7 @@ namespace pcl
   {
     public:
       typedef shared_ptr<BilateralUpsampling<PointInT, PointOutT> > Ptr;
-      typedef shared_ptr<const BilateralUpsampling<PointInT, PointOutT> > ConstPtr;
+      using ConstPtr = shared_ptr<const BilateralUpsampling<PointInT, PointOutT> >;
```
[the remark this answers is not in the document]
Comment (gnawme) [reply surface/include/pcl/surface/bilateral_upsampling.h:67]: Don't know. There are quite a few unconverted `typedef` statements, I'll have to cross-reference their filenames with the `compile_commands.json` list.

**Comment by gnawme** [comment]
> There's still plenty left (split default).
> 
> You can catch them all using `git diff master | grep '^ += default'

I ran that regex in my IDE originally. Running this on the command line didn't display any others.

File: io/include/pcl/compression/color_coding.h (hunk 6 above)
Comment (kunaltyagi) [inline io/include/pcl/compression/color_coding.h:75]: 1

File: io/include/pcl/compression/entropy_range_coder.h [hunk 9]
```diff
@@ -69,14 +69,12 @@ namespace pcl
 
     /** \brief Empty constructor. */
     AdaptiveRangeCoder ()
-    {
-    }
+    = default;
```
Comment (kunaltyagi) [inline io/include/pcl/compression/entropy_range_coder.h:72]: 3 here

File: io/include/pcl/compression/octree_pointcloud_compression.h [hunk 10]
```diff
@@ -119,9 +119,8 @@ namespace pcl
 
         /** \brief Empty deconstructor. */
         
-        ~OctreePointCloudCompression ()
-        {
-        }
+        ~OctreePointCloudCompression () override
```
Comment (kunaltyagi) [inline io/include/pcl/compression/octree_pointcloud_compression.h:122]: this one

File: io/include/pcl/compression/organized_pointcloud_compression.h [hunk 11]
```diff
@@ -61,13 +61,11 @@ namespace pcl
 
         /** \brief Empty Constructor. */
         OrganizedPointCloudCompression ()
-        {
-        }
+        = default;
```
Comment (kunaltyagi) [inline io/include/pcl/compression/organized_pointcloud_compression.h:64]: 2 here

File: io/include/pcl/compression/point_coding.h [hunk 12]
```diff
@@ -69,8 +69,7 @@ class PointCoding
     /** \brief Empty class constructor. */
     virtual
     ~PointCoding ()
-    {
-    }
+    = default;
```
Comment (kunaltyagi) [inline io/include/pcl/compression/point_coding.h:72]: 1 here

File: outofcore/include/pcl/outofcore/cJSON.h [hunk 13]
```diff
@@ -44,23 +44,37 @@ extern "C"
 #define cJSON_IsReference 256
 
 /* The cJSON structure: */
-typedef struct cJSON {
+using cJSON = struct cJSON {
```
Comment (kunaltyagi) [inline outofcore/include/pcl/outofcore/cJSON.h:47]: Please undo the changes in this file. Simply using `struct cJSON` (by removing the typedef and cJSON at the end) will suffice for a C++ header. `using` is not valid C (which I think this file was originally in.

File: recognition/include/pcl/recognition/3rdparty/metslib/abstract-search.hh [hunk 14]
```diff
@@ -109,9 +109,9 @@ namespace mets {
     abstract_search& operator==(const abstract_search<move_manager_type>&);
 
     /// @brief Virtual destructor.
-    virtual 
-    ~abstract_search() 
-    { };
+    
+    ~abstract_search() override 
+    = default;;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/3rdparty/metslib/abstract-search.hh:114]: 2 here

Comment (kunaltyagi) [inline recognition/include/pcl/recognition/3rdparty/metslib/abstract-search.hh:114]: With an extra ;

File: recognition/include/pcl/recognition/3rdparty/metslib/simulated-annealing.hh [hunk 15]
```diff
@@ -51,12 +51,12 @@ namespace mets {
   public:
     /// @brief Constructor
     abstract_cooling_schedule() 
-    { }
+    = default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/3rdparty/metslib/simulated-annealing.hh:54]: 2 more here

File: recognition/include/pcl/recognition/3rdparty/metslib/tabu-search.hh [hunk 16]
```diff
@@ -130,7 +130,7 @@ namespace mets {
     /// @brief Virtual destructor
     virtual 
     ~tabu_list_chain() 
-    { } 
+    = default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/3rdparty/metslib/tabu-search.hh:133]: 2 here as well

File: recognition/include/pcl/recognition/3rdparty/metslib/termination-criteria.hh [hunk 17]
```diff
@@ -59,7 +59,7 @@ namespace mets {
     /// @brief Virtual destructor.
     virtual 
     ~termination_criteria_chain() 
-    { }
+    = default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/3rdparty/metslib/termination-criteria.hh:62]: just the one here

File: recognition/include/pcl/recognition/face_detection/rf_face_detector_trainer.h [hunk 18]
```diff
@@ -77,9 +77,7 @@ namespace pcl
       }
 
       virtual ~RFFaceDetectorTrainer()
-      {
-
-      }
+      = default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/face_detection/rf_face_detector_trainer.h:80]: 1 here

File: recognition/include/pcl/recognition/face_detection/rf_face_utils.h [hunk 19]
```diff
@@ -204,9 +204,8 @@ namespace pcl
         }
 
         /** \brief Destructor. */
-        ~PoseClassRegressionVarianceStatsEstimator()
-        {
-        }
+        ~PoseClassRegressionVarianceStatsEstimator() override
+        = default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/face_detection/rf_face_utils.h:208]: and here

File: recognition/include/pcl/recognition/linemod.h [hunk 20]
```diff
@@ -61,8 +61,7 @@ namespace pcl
 
       /** \brief Destructor. */
       virtual ~EnergyMaps () 
-      {
-      }
+      = default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/linemod.h:64]: 2 here

File: recognition/include/pcl/recognition/linemod/line_rgbd.h [hunk 21]
```diff
@@ -106,8 +106,7 @@ namespace pcl
 
       /** \brief Destructor */
       virtual ~LineRGBD ()
-      {
-      }
+      = default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/linemod/line_rgbd.h:109]: 1 here

File: recognition/include/pcl/recognition/ransac_based/bvh.h [hunk 22]
```diff
@@ -74,8 +74,7 @@ namespace pcl
             }
 
             virtual ~BoundedObject ()
-            {
-            }
+            = default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/ransac_based/bvh.h:77]: 1 here

File: recognition/include/pcl/recognition/ransac_based/hypothesis.h [hunk 23]
```diff
@@ -65,7 +65,7 @@ namespace pcl
           memcpy (rigid_transform_, rigid_transform, 12*sizeof (float));
         }
 
-        virtual  ~HypothesisBase (){}
+        virtual  ~HypothesisBase ()= default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/ransac_based/hypothesis.h:68]: space before =, here and L98

**Review by kunaltyagi — CHANGES_REQUESTED** [review]
268/317 done

Almost all other files have similar issues (space or split  default)

File: io/include/pcl/compression/color_coding.h (hunk 6 above)
Comment (gnawme) [reply io/include/pcl/compression/color_coding.h]: Dang, I need my eyes checked...

File: recognition/include/pcl/recognition/3rdparty/metslib/simulated-annealing.hh (hunk 15 above)
Comment (gnawme) [reply recognition/include/pcl/recognition/3rdparty/metslib/simulated-annealing.hh:54]: I wonder why the hell some of these didn't show up in my find window?

File: recognition/include/pcl/recognition/ransac_based/hypothesis.h (hunk 23 above)
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/ransac_based/hypothesis.h:68]:
```suggestion
        virtual ~HypothesisBase () = default;
```

File: recognition/include/pcl/recognition/ransac_based/obj_rec_ransac.h [hunk 24]
```diff
@@ -104,7 +104,7 @@ namespace pcl
             {
               memcpy(this->rigid_transform_, rigid_transform, 12*sizeof (float));
             }
-            virtual ~Output (){}
+            virtual ~Output ()= default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/ransac_based/obj_rec_ransac.h:107]:
```suggestion
            virtual ~Output () = default;
```

File: recognition/include/pcl/recognition/ransac_based/obj_rec_ransac.h [hunk 25]
```diff
@@ -121,7 +121,7 @@ namespace pcl
               {
               }
 
-              virtual ~OrientedPointPair (){}
+              virtual ~OrientedPointPair ()= default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/ransac_based/obj_rec_ransac.h:124]:
```suggestion
              virtual ~OrientedPointPair () = default;
```

File: recognition/include/pcl/recognition/ransac_based/obj_rec_ransac.h [hunk 26]
```diff
@@ -130,8 +130,8 @@ namespace pcl
         class HypothesisCreator
         {
           public:
-            HypothesisCreator (){}
-            virtual ~HypothesisCreator (){}
+            HypothesisCreator ()= default;
+            virtual ~HypothesisCreator ()= default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/ransac_based/obj_rec_ransac.h:134]:
```suggestion
            HypothesisCreator () = default;
            virtual ~HypothesisCreator () = default;
```

File: surface/include/pcl/surface/mls.h [hunk 27]
```diff
@@ -262,7 +262,7 @@ namespace pcl
   {
     public:
       typedef shared_ptr<MovingLeastSquares<PointInT, PointOutT> > Ptr;
-      typedef shared_ptr<const MovingLeastSquares<PointInT, PointOutT> > ConstPtr;
+      using ConstPtr = shared_ptr<const MovingLeastSquares<PointInT, PointOutT> >;
```
Comment (kunaltyagi) [inline surface/include/pcl/surface/mls.h:265]: typedef above this line

File: recognition/include/pcl/recognition/ransac_based/orr_graph.h [hunk 28]
```diff
@@ -69,7 +69,7 @@ namespace pcl
               state_(UNDEF)
             {}
 
-            virtual ~Node (){}
+            virtual ~Node ()= default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/ransac_based/orr_graph.h:72]: space before =

File: recognition/include/pcl/recognition/ransac_based/orr_graph.h [hunk 29]
```diff
@@ -124,7 +124,7 @@ namespace pcl
         };
 
       public:
-        ORRGraph (){}
+        ORRGraph ()= default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/ransac_based/orr_graph.h:127]: and here

File: recognition/include/pcl/recognition/ransac_based/orr_octree.h [hunk 30]
```diff
@@ -89,7 +89,7 @@ namespace pcl
                   n_[0] = n_[1] = n_[2] = p_[0] = p_[1] = p_[2] = 0.0f;
                 }
 
-                virtual~ Data (){}
+                virtual~ Data ()= default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/ransac_based/orr_octree.h:92]: here

File: recognition/include/pcl/recognition/ransac_based/rigid_transform_space.h [hunk 31]
```diff
@@ -138,7 +138,7 @@ namespace pcl
         };// class Entry
 
       public:
-        RotationSpaceCell (){}
+        RotationSpaceCell ()= default;
```
Comment (kunaltyagi) [inline recognition/include/pcl/recognition/ransac_based/rigid_transform_space.h:141]: and this file

**Review by kunaltyagi — CHANGES_REQUESTED** [review]
310/316 files reviewed

**Review by kunaltyagi — APPROVED** [review]
LGTM

Thanks for bearing with us 😸

**Comment by gnawme** [comment]
> LGTM
> 
> Thanks for bearing with us

Happy to help.

**Review by mvieth — APPROVED** [review]
Thank you, and sorry for the delay
